README.md
# MSDFGL - OpenGL-accelerated implementation of the multi-channel signed distance-field algorithm.

![Demo](img/scale.png)

The MSDF algorithm implemented in this library was created by Viktor Chlumský (https://github.com/Chlumsky/msdfgen). Details about the implementation and the benefits of MSDF can be read from his repository or his thesis. MSDFGL is a GPU-accelerated implementation of his algorithm, and it was written from scratch in C.

The code had to go through quite a bit of modifications to make it runnable on the GPU:
1. It's no longer object-oriented
2. It now runs with constant memory (which is a requirement in shaders)
3. It does not use pointers
4. Dropped support for cubic segments (for simplicity's sake, this can probably be added later)

For now tested only with OpenGL ES 3.2 on Wayland EGL, and OpenGL 3.3 Core with glfw3, on Linux with Wayland, with Intel UHD Graphics 620, and on macOS Mojave on Macbook Pro 2015 (Iris Pro), and on Windows 10 with Intel HD Graphics 515. (Funny, I seem to currently own only Intel graphics cards...)

## Performance
The following plot shows performance comparisons to `msdfgen`. The benchmark consists of generating an MSDF texture from ASCII (0 - 127) characters for font `DejaVu Sans`. It was performed with a Core i7 8550m and Intel UHD Graphics 620, on Debian Linux. Each time was calculated by taking an average of a 100 executions, and canceling out the time for glfw to create and destroy the OpenGL context. 

> As the shader is quite complex, it's compilation takes close to a second. However, after one compilation, the user can create as many font textures as they want. Time to compile the shader is not included in the following comparisons.

The plot on the right has the exact same data, just with a logarithmic y-axis.

![Benchmarks](img/benchmarks.png)

> Values 8.0000 and 31.0000 for 8.0x msdfgen were extrapolated, as the CPU implementation scales quadratically (+ I didn't want to torture my Dell's CPU anymore...)

There seems to be some caching going on with the Intel driver on Linux, as the shader compilation seems to be instantaneous if the binary had been run before. First I thought that the resulting atlas texture was being cached, but generating the the texture was still fast to a font that msdfgl had never seen before. The caching behavior was not seen on macOS.

With `msdfgen` the measurement did not include the time to transfer the resulted bitmaps to a texture, whereas with `msdfgl` after the execution the font was ready to be rendered with.


## Implementation
The highly parallelizable part of MSDF algorithm has been moved to run on the GPU (the part of msdfgen which is executed per each pixel of the bitmap).

A loaded msdfgl font has two textures:
- Atlas texture - 2D RGBA texture containing all the generated MSDF bitmaps
- Index texture - 1D FLOAT texture buffer containing the coordinates and dimensions of each glyph on the atlas texture (there is also information about the bearing of the glyph so that we do not have to store the bitmap all the way from the origin, only from where the glyph actually starts).

Both the atlas and index textures grow as more glyphs are rendered. The user can render all the desired glyphs in bulk, or render them dynamically as new glyphs are introduced. -- Or a combination of those, for example render ASCII characters at the beginning, and then all the other characters as they are used. Rendering multiple characters at once yields better performance as we don't have to perform multiple copy-to-GPU operations nor to re-bind the buffers and shader in between.

![Implementation](img/diagram.png)

## Installation:

### Linux and macOS
The lib will be installed in `/usr/local/lib`, and the header into `/usr/local/include`.
```sh
mkdir build
cd build
cmake ..
make
sudo make install
```

### Windows
I tested that the compilation can be done and the example works on Windows with 32-bit MSVC 2019.

In `CMakeLists.txt` modify variables `FREETYPE_DIR` and `GLFW3_LIB` to point to an existing freetype2 install and to glfw3 library file. 32-bit and 64-bit freetype 2.10 are included in the repository for convenience.

```sh
mkdir build
cd build
cmake -G "NMake Makefiles" ..
nmake
```

## Usage:

An example is provided in `example`-directory. It can be compiled (requires glfw):
```sh
# In the cmake build directory
make msdfgldemo  # nmake instead of make on Windows
```
...and run:
```sh
./msdfgldemo /path/to/font/file.ttf "Hello, MSDFGL!"
```

### Usage as a library:
```C
#include <msdfgl.h>

msdfgl_context_t context;
context = msdfgl_create_context(NULL);  /* NULL defaults to "330 core" */

msdfgl_font_t font;
font = msdfgl_load_font(context, "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf",
                        4.0, 2.0, NULL); /* range, scale, atlas (NULL creates a new one) */

/* Loads characters 0-128 onto the textures. This is where all the GPU cycles went. */
msdfgl_generate_ascii(font);

/*            x    y           size  color       4x4 projection-matrix  flags */
msdfgl_printf(0.0, 20.0, font, 18.0, 0xffffffff, (GLfloat *)projection, MSDFGL_KERNING,
              "Hello, MSDFGL v%s!", MSDFGL_VERSION);

/* Cleanup */
msdfgl_destroy_font(font);
msdfgl_destroy_context(context);
```
Then, link your binary against msdfgl:
```sh
gcc ... -lmsdfgl
```

The library includes two shaders:
- Generator shader - heavy lifting, generates the MSDF bitmaps.
- Render shader - renders crisp text from the generated textures.


## TODO:
- Edge-coloring for teardrop with 1 or 2 segments
- Support for cubic segments (needed for OpenType fonts)
- Unit testing


CHANGELOG.md
# Changelog

## [0.2] - 2019-06-23
### Changes:
- Fixed noise from the atlas texture
- Moved all printf-commands behind one API call
- Bugfixes, error handling improvements and cleanup
- Support using the same texture with multiple fonts
- Static library on Linux/macOS
- Render only one "empty box" for control characters.
### Known issues:
- No support for cubic segments yet
- Color mapping for certain shapes is incomplete
- Rendering calls use a geomety shader
- Small errors in shapes, such as letters `q` and `k` with `LiberationSerif-Regular.ttf`. Visible only with a really large font size.
- DLL build on windows not working


include/msdfgl.h
#ifndef MSDFGL_H
#define MSDFGL_H

/**
 * OpenGL implementation for multi-channel signed distance field generator
 * -----------------------------------------------------------------------
 *
 * msdfgl              Henrik Nyman,    (c) 2019 -
 * msdfgen             Viktor Chlumsky, (c) 2014 - 2019
 *
 * The technique used to generate multi-channel distance fields in this code has
 * been developed by Viktor Chlumsky in 2014 for his master's thesis, "Shape
 * Decomposition for Multi-Channel Distance Fields". It provides improved
 * quality of sharp corners in glyphs and other 2D shapes in comparison to
 * monochrome distance fields. To reconstruct an image of the shape, apply the
 * median of three operation on the triplet of sampled distance field values.
 *
 * MSDFGL provides a version of that algorithm that runs partially on the GPU,
 * and a higher-level API handling font atlases and textures.
 */

#ifdef _WIN32
#ifdef MSDFGL_EXPORTS
#define MSDFGL_EXPORT __declspec(dllexport)
#else
#define MSDFGL_EXPORT __declspec(dllimport)
#endif

#include <stdint.h>
#include <windows.h>

#else
#define MSDFGL_EXPORT
#endif

#include <ft2build.h>
#include FT_FREETYPE_H
#include FT_OUTLINE_H

#ifdef __APPLE__
#include <OpenGL/gl3.h>
#else
#include <GL/gl.h>
#endif

#define MSDFGL_VERSION 0.2

#ifdef __cplusplus
extern "C" {
#endif

typedef struct _msdfgl_context *msdfgl_context_t;
typedef struct _msdfgl_font *msdfgl_font_t;
typedef struct _msdfgl_atlas *msdfgl_atlas_t;

/**
 * Compile shaders and configure uniforms.
 *
 * version-parameter defines with which GLSL version the shaders will be
 * compiled. Supported versions are "330", "330 core" and "320 es".
 * giving NULL uses the default ("330 core").
 *
 * Returns a new MSDF GL context, or NULL if creating the context failed.
 */
MSDFGL_EXPORT msdfgl_context_t msdfgl_create_context(const char *version);

/**
 * Release resources allocated by `msdfgl_crate_context`.
 */
MSDFGL_EXPORT void msdfgl_destroy_context(msdfgl_context_t ctx);

/**
 * Serialized format of a glyph.
 */
typedef struct _msdfgl_glyph {
    /**
     * X and Y coordinates in in the projection coordinates.
     */
    GLfloat x;
    GLfloat y;

    /**
     * The color of the character in 0xRRGGBBAA format.
     */
    GLuint color;

    /**
     * Unicode code point of the character.
     */
    GLint key;

    /**
     * Font size to use for rendering of this character.
     */
    GLfloat size;

    /**
     * Y offset (for e.g. subscripts and superscripts).
     */
    GLfloat offset;

    /**
     * The amount of "lean" on the character. Positive leans to the right,
     * negative leans to the left. Skew can create /italics/ effect without
     * loading a separate font atlas.
     */
    GLfloat skew;

    /**
     * The "boldness" of the character. 0.5 is normal strength, lower is thinner
     * and higher is thicker. Strength can create *bold* effect without loading
     * a separate font atlas.
     */
    GLfloat strength;

} msdfgl_glyph_t;

/**
 * Allocate textures for a font atlas.
 */
MSDFGL_EXPORT msdfgl_atlas_t msdfgl_create_atlas(msdfgl_context_t ctx, int texture_width,
                                                 int padding);

/**
 * Release resources of a font atlas.
 */
MSDFGL_EXPORT void msdfgl_destroy_atlas(msdfgl_atlas_t atlas);

/**
 * Load font from a font file and generate textures and buffers for it.
 */
MSDFGL_EXPORT msdfgl_font_t msdfgl_load_font(msdfgl_context_t ctx, const char *font_name,
                                             float range, float scale,
                                             msdfgl_atlas_t atlas);

/**
 * Get vertical advance of the font with the given size.
 */
MSDFGL_EXPORT float msdfgl_vertical_advance(msdfgl_font_t font, float size);

/**
 * Release resources allocated by `msdfgl_load_font`.
 */
MSDFGL_EXPORT void msdfgl_destroy_font(msdfgl_font_t font);

/**
 * Render a single glyph onto the MSDF atlas. Intented use case is to generate
 * the bitmaps on-demand as the characters are appearing.
 */
MSDFGL_EXPORT int msdfgl_generate_glyph(msdfgl_font_t font, int32_t char_code);

/**
 * Render a range of glyphs onto the MSDF atlas. The range is inclusive. Intended
 * use case is to initialize the atlas in the beginning with e.g. ASCII characters.
 */
MSDFGL_EXPORT int msdfgl_generate_glyphs(msdfgl_font_t font, int32_t start, int32_t end);

/**
 * Render arbitrary character codes in bulk.
 */
MSDFGL_EXPORT int msdfgl_generate_glyph_list(msdfgl_font_t font, int32_t *list, size_t n);

/**
 * Shortcuts for common generators.
 */
#define msdfgl_generate_ascii(font) msdfgl_generate_glyphs(font, 0, 127)
#define msdfgl_generate_ascii_ext(font) msdfgl_generate_glyphs(font, 0, 255)

/**
 * Render a list of glyphs.
 */
MSDFGL_EXPORT void msdfgl_render(msdfgl_font_t font, msdfgl_glyph_t *glyphs, int n,
                                 GLfloat *projection);

/**
 * Printf options.
 */
enum msdfgl_printf_flags {

    /**
     * Use FreeType kerning if it is available for the font.
     */
    MSDFGL_KERNING = 0x01,

    /**
     * Render wide character arrays. Give the fmt-argment as wchar_t *.
     */
    MSDFGL_WCHAR = 0x02,

    /**
     * Draw text vertically instead of horizontally.
     */
    MSDFGL_VERTICAL = 0x04,
};

/**
 * Print a formatted string on currently active framebuffer.
 *
 * x - x coordinate of the starting glyph.
 * y - y coordinate of the starting glyph (note that this points to the origin
 *     of the glyph, which is usually at bottom left).
 * font - font to use for rendering.
 * size - font size in points (fractional sizes are allowed).
 * color - color of the text (0xrrggbbaa).
 * projection - 4x4 projection matrix.
 * flags - drawing options, see `enum msdfgl_printf_flags`
 * fmt - char * (or wchar_t *) format string.
 * ... - substitution variables for the format string.
 *
 * Returns the x (y if vertical drawing is enabled) position of the glyph that
 * would follow the rendered ones.
 */
MSDFGL_EXPORT float msdfgl_printf(float x, float y, msdfgl_font_t font, float size,
                                  int32_t color, GLfloat *projection,
                                  enum msdfgl_printf_flags flags, const void *fmt, ...);

/**
 * Set the DPI for the current session. Following draw calls will use the new DPI.
 * The DPI value is a vector, which allows for rendering text with a monitor which
 * has non-square pixels.
 * The default DPI is (72, 72).
 */
MSDFGL_EXPORT void msdfgl_set_dpi(msdfgl_context_t context, float horizontal,
                                  float vertical);

/* Plumbing commands. In case you want to build your own renderer. */
/**
 * Generates an orthographic projection (similar to glm's ortho).
 */
MSDFGL_EXPORT void _msdfgl_ortho(GLfloat left, GLfloat right, GLfloat bottom, GLfloat top,
                                 GLfloat nearVal, GLfloat farVal, GLfloat dest[][4]);

/**
 * Get the atlas texture of the given font.
 */
MSDFGL_EXPORT GLuint _msdfgl_atlas_texture(msdfgl_font_t font);
/**
 * Get the index texture of the given font.
 */
MSDFGL_EXPORT GLuint _msdfgl_index_texture(msdfgl_font_t font);
#ifdef __cplusplus
}
#endif

#endif /* MSDFGL_H */
